### PR TITLE
add #12 page faults lecture

### PR DESCRIPTION
**Pull request by luk3b**

### CODE REVIEW

File: betriebssysteme/12-page-faults.md [hunk 1]
```diff
@@ -4,6 +4,279 @@
...
+
+The MMU is a hardware device which maps virtual to physical addresses. User programs always use virtual addresses and never see, nor know about, physical addresses.
+
+![](img/09-mmu.png)
+
+## Revision: Paging
+
+![](img/10-paging.png)
+
+A so called **present bit** in the page table indicates whether a virtual **page** is currently mapped to physical memory, which the MMU automaticly translates (if valid). If a virtual address is used and currently _not_ mapped to physical memory, the MMU calls the OS to bring in the data. This is called a **page fault**.
```
Comment (pbrinkmeier) [inline betriebssysteme/12-page-faults.md]: so-called*

and please put each sentence on an extra line, for cleaner diffs

File: betriebssysteme/12-page-faults.md [hunk 2]
```diff
@@ -4,6 +4,279 @@
...
+    =           200ns + p * 7.999.800ns
+```
+
+If only one access out of 1.000 causes a page fault, the EAT will calculate to 8.2ms which is a slowdown by a factor of 40!
+
+### Challenges
+
+Page faults introduce a set of challenges, including the following questions:
+
+- What to eject? _(=> frame allocation)_
```
Comment (pbrinkmeier) [inline betriebssysteme/12-page-faults.md]: Please use ⇒

File: betriebssysteme/12-page-faults.md [hunk 3]
```diff
@@ -4,6 +4,279 @@
...
+
+To resume a process after a page fault, some extra information is necessary. The hardware therefore provides info about the page fault. On Intel machines for example, the ```%cr2``` register contains the faulting virtual address.
+
+The OS then needs to figure out the context of the fault, this mainly includes identifying the faulting instruction:
+
+- read or write?
+- instruction fetch?
+- user access to kernel?
+
+**Idempotent instructions can just re-executed**, however more complex instructions must be re-started too.
```
Comment (pbrinkmeier) [inline betriebssysteme/12-page-faults.md]: can just be*

File: betriebssysteme/12-page-faults.md [hunk 4]
```diff
@@ -4,6 +4,279 @@
...
+
+![](img/12-shared-memory.png)
+
+#### Memory-Mapped Files
+
+![](img/12-memory-mapped-files.png)
+
+Memory-mapped file I/O allows file I/O to be treated as routine memory access by mapping a disk block to a page. Initially, a page-sized portion of the file is read into a physical page. Subsequent reads from/writes to the file are treated as ordinary memory accesses.
+
+This simplifies file access because ```read()``` and ```write()``` system calls aren't necessary. Also, this allows several processes to map the same file.
```
Comment (pbrinkmeier) [inline betriebssysteme/12-page-faults.md]: one backtick for inline code :)

File: betriebssysteme/12-page-faults.md [hunk 5]
```diff
@@ -4,6 +4,279 @@
...
+
+Ideally you would replace the page that is going to be referenced furthest in the future, however it's impossible to reliably predict the future.
+
+The first idea is to predict the future from the past by recording page references from the past and extrapolate them into the future. However it's very expensive to maintain an ordered list of all page references at runtime.
+
+Therefore the second idea is to sacrifice precision for speed and let the MMU set the **reference bit** in the respective page table entry _every time a page is referenced_. A timer then scans all page tabe entries for reference bits.
+
+### Page Fault Frequency Allocation Scheme
+
+Establish a “acceptable” page fault rate with an upper and lower bound. If the actual rate is too low, give frames to other processes. Otherwise, allocate more frames to the process.
```
Comment (pbrinkmeier) [inline betriebssysteme/12-page-faults.md]: an*

**Comment by luk3b** [comment]
:+1:

**Pull request merged**